src/manager/manager.go

```diff
@@ -0,0 +1,230 @@
+        IncludeFolderEnd()
+
+        // IncludeFile reads the given file, adds it to the file, and creates the Metadata.
+        //
+        // parameter (fn)       : name of the file to be read
+        // paramter (basedir)   : name of the current directory relative to root
+        // return               : new offset into the image file
+        IncludeFile(fn string, basedir string) (int64, error)
+
+        // WriterHeader writes the Metadata for the imagefile to the end of the image file.
+        // The location of the beginning of the header is written at the very end as an int64
+        WriteHeader() error
+}
+
+// FileMetadata holds information for the location of a file in the image file
+type FileMetadata struct {
+        // Begin indicates the beginning of a file (pointer) in the file
+        Begin int64
+
+        // End indicates the ending of a file (pointer) in the file
+        End int64
+
+        // Name indicates the name of a specific file in the file
+        Name string
+
+        // If the file is a symlink, this entry is used for link info
+        Link string
+
+        // Type indicated the type of a specific file (dir, symlink or regular file)
+        Type fileType
+}
+
+// Manager is the main driver of creating the image file. It writes the data and stores Metadata.
+type ZarManager struct {
+        // PageAlign indicates whether files will be aligned at page boundaries
+        PageAlign bool
+
+        // The FileWriter for this zar image
+        Writer writer.FileWriter
+
+        // Metadata is a list of FileMetadata structs indicating start and end of directories and files
+        Metadata []FileMetadata
+}
+
+// WalkDir implemented Manager.WalkDir
+func (z *ZarManager) WalkDir(dir string, foldername string, root bool) {
+        // root dir not marked as directory
+        if !root {
+                fmt.Printf("including folder: %v, name: %v\n", dir, foldername)
+                z.IncludeFolderBegin(foldername)
+        }
+
+        // Retrieve all files in current directory
+        files, err := ioutil.ReadDir(dir)
+        if err != nil {
+                log.Fatalf("walk dir unknown err when processing dir %v", dir)
+        }
+
+        var dirs []string
+
+        // Process each file in the directory
+        for _, file := range files {
+                name := file.Name()
+                symlink := file.Mode() & os.ModeSymlink != 0
+                file_path := path.Join(dir, name)
+
+                if symlink {
+                        // Symbolic link is an indirection, thus read and include
+                        fmt.Printf("%v is symlink.", file_path)
+                        real_dest, err := os.Readlink(file_path)
+                        if err != nil {
+                                log.Fatalf("error. Can't read symlink file. %v", real_dest)
+                        }
+                        // TODO: Can we replace with file redirecting to here? Could eliminate symbolic links
```

Comment:
Do you mean the feature that replace symbol link to hard link?

I think we can still add symbol link here. Remember the symbol link location. Later we replaced all symbol link metadata to hard link metadata. Why? Because we don't know some symbol link files such as "../1.txt" has already written to the image or not. Maybe it hasn't been "walked" so we can't generate offset information for the "hard link".

Reply:
Yes, I think that is a good idea. I think that having a global structure to save soft link info until the hard link is written would be good (would eliminate all symbolic links in image file).

Perhaps, if we find that locality matters in the image file, we could add the actual file that the symbolic link refers to at this location (since it is referenced here). This would lead to duplication, but could get rid of both symbolic and hard links in image file (I think).